src/apps/lwaftr/fragmentv4.lua

```diff
@@ -0,0 +1,289 @@
+--     an ICMP Datagram Too Big (Type 3, Code 4) packet back to the sender.
+--
+function fragment_ipv4(ipv4_pkt, mtu)
+   if ipv4_pkt.length <= mtu then
+      return FRAGMENT_UNNEEDED, ipv4_pkt
+   end
+
+   -- Discard packets with the DF (dont't fragment) flag set
+   do
+		local flags_and_frag_offset = C.ntohs(rd16(ipv4_pkt.data + flags_and_frag_offset_offset))
+		if band(flags_and_frag_offset, flag_dont_fragment_mask) ~= 0 then
+			local icmp_pkt = packet.allocate()
+			return FRAGMENT_FORBIDDEN, nil
+		end
+	end
+
+   local ihl = get_ihl(ipv4_pkt)
+   local header_size = constants.ethernet_header_size + ihl
+   local payload_size = ipv4_pkt.length - header_size
+   -- Payload bytes per packet must be a multiple of 8
+   local payload_bytes_per_packet = band(mtu - header_size, 0xFFF8)
+   local total_length_per_packet = payload_bytes_per_packet + ihl
+   local num_packets = ceil(payload_size / payload_bytes_per_packet)
+
+   local pkts = { ipv4_pkt }
+
+   wr16(ipv4_pkt.data + frag_id_offset, C.htons(fresh_frag_id()))
+   wr16(ipv4_pkt.data + total_length_offset, C.htons(total_length_per_packet))
+   wr16(ipv4_pkt.data + flags_and_frag_offset_offset, C.htons(flag_more_fragments_mask))
+
+   local raw_frag_offset = payload_bytes_per_packet
+
+   for i = 2, num_packets - 1 do
+      local frag_pkt = packet.allocate()
+      ffi.copy(frag_pkt.data, ipv4_pkt.data, header_size)
+      ffi.copy(frag_pkt.data + header_size,
+               ipv4_pkt.data + header_size + raw_frag_offset,
+               payload_bytes_per_packet)
+		wr16(frag_pkt.data + flags_and_frag_offset_offset,
+           C.htons(bor(flag_more_fragments_mask,
+                       band(frag_offset_field_mask, raw_frag_offset / 8))))
+      frag_pkt.length = header_size + payload_bytes_per_packet
+      raw_frag_offset = raw_frag_offset + payload_bytes_per_packet
+      pkts[i] = frag_pkt
+   end
+
+   -- Last packet
+   local last_pkt = packet.allocate()
+   local last_payload_len = payload_size % payload_bytes_per_packet
+   ffi.copy(last_pkt.data, ipv4_pkt.data, header_size)
+   ffi.copy(last_pkt.data + header_size,
+            ipv4_pkt.data + header_size + raw_frag_offset,
+            last_payload_len)
+   wr16(last_pkt.data + flags_and_frag_offset_offset,
+        C.htons(band(frag_offset_field_mask, raw_frag_offset / 8)))
+   wr16(last_pkt.data + total_length_offset, C.htons(last_payload_len + ihl))
+   last_pkt.length = header_size + last_payload_len
+   pkts[num_packets] = last_pkt
+
+   -- Truncate the original packet
+   ipv4_pkt.length = header_size + payload_bytes_per_packet
+
+   return FRAGMENT_OK, pkts
+end
+
+
+function selftest()
+   print("selftest: lwaftr.fragmentv4.fragment_ipv4")
+
+   local eth_proto = require("lib.protocol.ethernet")
+   local ip4_proto = require("lib.protocol.ipv4")
+
+   -- Makes an IPv4 packet, with Ethernet framing, with a given payload size
+   local function make_ipv4_packet(payload_size)
+      local pkt = packet.allocate()
```

Comment:
It's also possible to create a packet out of raw data. There's an example of this in ```lib/protocol/tcp.lua```:

```lua
   local p = packet.from_string(lib.hexundump([[
      52:54:00:02:02:02 52:54:00:01:01:01 08 00 45 00
      00 34 59 1a 40 00 40 06 00 00 c0 a8 14 a9 6b 15
      f0 b4 de 0b 01 bb e7 db 57 bc 91 cd 18 32 80 10
      05 9f 00 00 00 00 01 01 08 0a 06 0c 5c bd fa 4a
      e1 65
   ]], 66))
```
Maybe it was possible to use something like that here. The code as it is now, it's OK to me.